optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/bavet/uni/BavetGroupBridgeUniNode.java

```diff
@@ -0,0 +1,114 @@
+            Function<A, GroupKey_> groupKeyMapping, UniConstraintCollector<A, ResultContainer_, Result_> collector,
+            BavetGroupBiNode<GroupKey_, ResultContainer_, Result_> biNode) {
+        super(session, nodeOrder);
+        this.parentNode = parentNode;
+        this.groupKeyMapping = groupKeyMapping;
+        this.collector = collector;
+        this.biNode = biNode;
+        tupleMap = new HashMap<>();
+    }
+
+    @Override
+    public BavetGroupBridgeUniTuple<A, GroupKey_, ResultContainer_, Result_> createTuple(BavetAbstractUniTuple<A> parentTuple) {
+        return new BavetGroupBridgeUniTuple<>(this, parentTuple);
+    }
+
+    public void refresh(BavetGroupBridgeUniTuple<A, GroupKey_, ResultContainer_, Result_> tuple) {
+        if (tuple.getChildTuple() != null) {
+            BavetGroupBiTuple<GroupKey_, ResultContainer_, Result_> childTuple = tuple.getChildTuple();
+            GroupKey_ oldGroupKey = childTuple.getGroupKey();
+            int parentCount = childTuple.decreaseParentCount();
+            tuple.getUndoAccumulator().run();
+            childTuple.clearResult();
+            tuple.setChildTuple(null);
+            tuple.setUndoAccumulator(null);
+            if (parentCount == 0) {
+                // Clean up tupleMap
+                tupleMap.remove(oldGroupKey);
+                session.transitionTuple(childTuple, BavetTupleState.DYING);
+            } else {
+                session.transitionTuple(childTuple, BavetTupleState.UPDATING);
+            }
+        }
+        if (tuple.isActive()) {
+            A a = tuple.getFactA();
+            GroupKey_ groupKey = groupKeyMapping.apply(a);
+            BavetGroupBiTuple<GroupKey_, ResultContainer_, Result_> childTuple = tupleMap.computeIfAbsent(groupKey,
+                    k -> biNode.createTuple(groupKey, collector.supplier().get()));
+            int parentCount = childTuple.increaseParentCount();
+
+            Runnable undoAccumulator = collector.accumulator().apply(childTuple.getResultContainer(), a);
+            tuple.setUndoAccumulator(undoAccumulator);
+            childTuple.clearResult();
+            tuple.setChildTuple(childTuple);
+            if (parentCount == 1) {
+                session.transitionTuple(childTuple, BavetTupleState.CREATING);
+            } else {
+                // It might have just been created by an earlier tuple in the same nodeOrder
+                if (childTuple.getState() != BavetTupleState.CREATING) {
+                    session.transitionTuple(childTuple, BavetTupleState.UPDATING);
+                }
+            }
+        }
+        tuple.refreshed();
+    }
+
+    @Override
+    public String toString() {
+        return "GroupBridge()";
+    }
+
+    // ************************************************************************
+    // Getters/setters
+    // ************************************************************************
+
+    private class GroupKeyData {
```

Comment:
This is a weird class. Doesn't seem to do anything. Holds data and possibly lets them be mutated from the outside. If this is actually being used anywhere, it looks like a `Map.Entry`.

Reply:
Oops - should have been deleted already :) Done